brave/src/main/java/brave/internal/propagation/CorrelationFieldScopeDecorator.java

```diff
@@ -13,85 +13,176 @@
+
+    ExtraFieldUpdater(CorrelationFieldScopeDecorator decorator, String name) {
+      super(decorator, name);
+      this.propagationType = InternalPropagation.instance.extraPropagationFieldsType();
+    }
+
+    @Override boolean update(TraceContext context, @Nullable String previous) {
+      String current = PropagationFields.get(context, field, propagationType);
+      return updateNullable(previous, current);
     }
+  }
 
-    String spanId = currentSpan.spanIdString();
-    if (!spanId.equals(previousSpanId)) put("spanId", spanId);
+  static final class TraceIdUpdater extends Updater {
+    TraceIdUpdater(CorrelationFieldScopeDecorator decorator) {
+      super(decorator, "traceId");
+    }
 
-    Boolean sampled = currentSpan.sampled();
-    if (sampled == null) {
-      remove("sampled");
-    } else {
-      String sampledString = sampled.toString();
-      boolean sameSampled = sampledString.equals(previousSampled);
-      if (!sameSampled) put("sampled", sampledString);
+    @Override boolean update(TraceContext context, @Nullable String previous) {
+      return update(previous, context.traceIdString());
+    }
+  }
+
+  static final class ParentSpanIdUpdater extends Updater {
+    ParentSpanIdUpdater(CorrelationFieldScopeDecorator decorator) {
+      super(decorator, "parentId");
+    }
+
+    @Override boolean update(TraceContext context, @Nullable String previous) {
+      return updateNullable(previous, context.parentIdString());
+    }
+  }
+
+  static final class SpanIdUpdater extends Updater {
+    SpanIdUpdater(CorrelationFieldScopeDecorator decorator) {
+      super(decorator, "spanId");
+    }
+
+    @Override boolean update(TraceContext context, @Nullable String previous) {
+      return update(previous, context.spanIdString());
+    }
+  }
+
+  static final class SampledUpdater extends Updater {
+    SampledUpdater(CorrelationFieldScopeDecorator decorator) {
+      super(decorator, "sampled");
+    }
+
+    @Override boolean update(TraceContext context, @Nullable String previous) {
+      Boolean sampled = context.sampled();
+      return updateNullable(previous, sampled != null ? sampled.toString() : null);
+    }
+  }
+
+  static abstract class Updater {
+    final CorrelationFieldScopeDecorator decorator;
+    final String field;
+
+    Updater(CorrelationFieldScopeDecorator decorator, String field) {
+      this.decorator = decorator;
+      this.field = field;
+    }
+
+    /** Returns true if there was a change to the correlation field. */
+    abstract boolean update(TraceContext context, @Nullable String previous);
+
+    boolean update(@Nullable String previous, String current) {
+      if (!current.equals(previous)) {
+        decorator.put(field, current);
+        return true;
+      }
+      return false;
+    }
+
+    boolean updateNullable(@Nullable String previous, @Nullable String current) {
+      if (current != null) return update(previous, current);
```

Comment:
Small nit this is probably easier to read if the if/else pattern is consistent, currently this line seems particularly special for some reason.

```java
if (current != null) {
  return update(previous, current);
} else if (previous != null) {
  decorator.remove(field);
  return true;
} else {
  return false;
}
```

Could remove the elses if desired.